File: 2024-1/Week_6/김민규/p1.py
import sys

def main():
    input = sys.stdin.read
    data = input().strip().split('\n')
    
    # 첫 번째 줄에서 N, S, M을 읽어옵니다.
    N, S, M = map(int, data[0].split())
    
    # 두 번째 줄에서 볼륨 조절 리스트 V를 읽어옵니다.
    V = list(map(int, data[1].split()))
    

    dp = [[False] * (M+1) for _ in range(N+1)]
    dp[0][S] = True

    for i in range(1, N+1):
        for j in range(M+1):
            if dp[i-1][j]:
                if j+V[i-1] <= M:
                    dp[i][j+V[i-1]] = True
                if j-V[i-1] >= 0:
                    dp[i][j-V[i-1]] = True


                                
    max_volume = -1
    for j in range(M+1):
        if dp[N][j]:
            max_volume = max(max_volume, j)

    print(max_volume)

if __name__ == "__main__":
    main()

File: 2024-1/Week_6/김민규/p1.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];

rl.on('line', function (line) {
    input.push(line.trim());
}).on('close', function () {
    const [N, S, M] = input[0].split(' ').map(Number); // 곡의 개수 n 시작볼륨 s 최대볼륨 m
    const V = input[1].split(' ').map(Number);

    const dp = Array.from({length: N+1}, ()=> Array(M+1).fill(false));
    
    dp[0][S] = true;

    for(let i = 1; i <= N; i++){
        for(let j = 0; j <= M; j++){
            if(dp[i-1][j]){
                if(j+V[i-1] <= M) dp[i][j+V[i-1]] = true;
                if(j-V[i-1] >= 0) dp[i][j-V[i-1]] = true;
            }
        }
    }

    let maxVolume = -1;
    
    for(let j = 0; j <= M; j++){
        if(dp[N][j]) maxVolume = Math.max(maxVolume, j);
    }

    console.log(maxVolume);
});